### PR TITLE
chore: add generateCallbackUri() method to utils

### PR DESCRIPTION
**Pull request by darcyYe**
## Summary
add method of generateCallbackUri() to utils

## Linear Issue Reference
[Linear LOG-238](https://linear.app/silverhand/issue/LOG-238/js-add-utils-to-generate-callback-url)

## Testing
add uni-tests to utils.test.ts and pass all the tests

### CODE REVIEW

File: packages/client/src/index.test.ts [hunk 1]
```diff
@@ -170,7 +171,11 @@ describe('LogtoClient', () => {
         storage,
       });
       await logto.loginWithRedirect(REDIRECT_URI, jest.fn());
-      expect(logto.isLoginRedirect(`${REDIRECT_URI}?code=${CODE}&state=${STATE}`)).toBeTruthy();
+      expect(
```
Comment (simeng-li) [inline packages/client/src/index.test.ts:174]: prefer avoid nested call stacks, just to be more readable
```js
const callbackUri = generateCallbackUri({ redirectUri: REDIRECT_URI, code: CODE, state: STATE });
expect(logto.isLoginRedirect(callbackUri)).toBeTruthy();
```

File: packages/client/src/utils.test.ts [hunk 2]
```diff
@@ -32,3 +32,54 @@ describe('decodeToken', () => {
     expect(() => decodeToken(JWT)).toThrowError(StructError);
   });
 });
+
+describe('generateCallbackUri', () => {
+  test('generate callback url with redirectUri', () => {
+    const redirectUri = 'http://localhost:3000';
```
Comment (simeng-li) [inline packages/client/src/utils.test.ts:38]: normally we could name using all uppercase to indicate it is a static constant. i.e.  `REDIRECT_URI`

File: packages/client/src/utils.test.ts [hunk 3]
```diff
@@ -32,3 +32,54 @@ describe('decodeToken', () => {
...
+    expect(generateCallbackUri({ redirectUri })).toEqual('http://localhost:3000');
+  });
+
+  test('generate callback url with redirectUri and code', () => {
+    const redirectUri = 'http://localhost:3000';
```
Comment (simeng-li) [inline packages/client/src/utils.test.ts:43]: Looks like those constants are shared cross all the test cases. We could move them to the upper scope, so no need to define multiple times.

File: packages/client/src/utils.ts [hunk 4]
```diff
@@ -68,3 +76,39 @@ export const createDefaultOnRedirect = () => {
     window.location.assign(url);
   };
 };
+
+export const generateCallbackUri = (uriInfo: UriInfo): string => {
```
Comment (simeng-li) [inline packages/client/src/utils.ts:80]: could use the destruct feature
`({ redirectUri, code, state, error, errorDescription }: UriInfo) => {} `

File: packages/client/src/utils.ts [hunk 5]
```diff
@@ -68,3 +76,39 @@ export const createDefaultOnRedirect = () => {
...
+  const callbackUriBase: string = uriInfo.redirectUri
+    .slice()
+    .concat(uriInfo.code || uriInfo.state || uriInfo.error || uriInfo.errorDescription ? '?' : '');
+
+  const callbackUriConcatCode: string = callbackUriBase.concat(
+    uriInfo.code ? `code=${uriInfo.code}` : ''
+  );
+
+  const callbackUriConcatState: string = callbackUriConcatCode.concat(
```
Comment (simeng-li) [inline packages/client/src/utils.ts:89]: we do have[ query-string](https://github.com/sindresorhus/query-string) added as a dependency. Could leverage the library to handle the search params and append it to the `redirectUri`.   

1. retrieve any original query params from the `redirectUri`
2. merge the query params with all the `non-empty` callback params
3. generate the new callbackUri

**Review by simeng-li — COMMENTED** [review]
Please check

File: packages/client/src/utils.test.ts [hunk 6]
```diff
@@ -32,3 +38,45 @@ describe('decodeToken', () => {
...
+
+describe('generateCallbackUri', () => {
+  test('generate callback url with redirectUri', () => {
+    const callbackUri = generateCallbackUri({ redirectUri: REDIRECT_URI });
+    expect(callbackUri).toEqual('http://localhost:3000');
+  });
+
+  test('generate callback url with redirectUri and code', () => {
+    const callbackUri = generateCallbackUri({ redirectUri: REDIRECT_URI, code: CODE });
+    expect(callbackUri).toEqual('http://localhost:3000?code=code1');
```
Comment (IceHe) [inline packages/client/src/utils.test.ts:50]: `${REDIRECT_URI}?code=${code1}`

File: packages/client/src/utils.test.ts [hunk 7]
```diff
@@ -32,3 +38,45 @@ describe('decodeToken', () => {
     expect(() => decodeToken(JWT)).toThrowError(StructError);
   });
 });
+
+describe('generateCallbackUri', () => {
+  test('generate callback url with redirectUri', () => {
+    const callbackUri = generateCallbackUri({ redirectUri: REDIRECT_URI });
+    expect(callbackUri).toEqual('http://localhost:3000');
```
Comment (IceHe) [inline packages/client/src/utils.test.ts:45]: replace 'http://localhost:3000' with REDIRECT_URI

File: packages/client/src/utils.test.ts [hunk 8]
```diff
@@ -32,3 +38,45 @@ describe('decodeToken', () => {
...
+
+  test('generate callback url with redirectUri, code, state and error', () => {
+    const callbackUri = generateCallbackUri({
+      redirectUri: REDIRECT_URI,
+      code: CODE,
+      state: STATE,
+      error: ERROR,
+    });
+    expect(callbackUri).toEqual(
+      'http://localhost:3000?code=code1&state=state1&error=invalid_request'
```
Comment (IceHe) [inline packages/client/src/utils.test.ts:66]: ditto

File: packages/client/src/utils.test.ts [hunk 9]
```diff
@@ -32,3 +38,45 @@ describe('decodeToken', () => {
...
+  test('generate callback url with full info', () => {
+    const callbackUri = generateCallbackUri({
+      redirectUri: REDIRECT_URI,
+      code: CODE,
+      state: STATE,
+      error: ERROR,
+      errorDescription: ERROR_DESCRIPTION,
+    });
+    expect(callbackUri).toEqual(
+      'http://localhost:3000?code=code1&state=state1&error=invalid_request&error_description=code_challenge%20must%20be%20a%20string%20with%20a%20minimum%20length%20of%2043%20characters'
```
Comment (IceHe) [inline packages/client/src/utils.test.ts:79]: ditto

File: packages/client/src/utils.ts [hunk 10]
```diff
@@ -22,6 +23,14 @@ const IDTokenSchema = s.type({
   at_hash: s.optional(s.string()),
 });
 
+interface UriInfo {
```
Comment (IceHe) [inline packages/client/src/utils.ts:26]: TBD: Better to rename to `CallbackUriParameters`?

**Review by IceHe — CHANGES_REQUESTED** [review]

File: packages/client/src/utils.ts (hunk 10 above)
Comment (darcyYe) [reply packages/client/src/utils.ts:26]: good point

**Review by IceHe — APPROVED** [review]
LGTM

**Pull request merged**